### ISSUE TITLE
Refine no-wait-for-multiple-assertions: Only forbid multiple assertions on the same async target

### ISSUE THREAD

**Issue by y-hsgw** [testing-library/eslint-plugin-testing-library#1084]
### What rule do you want to change?

no-wait-for-multiple-assertions

### Does this change cause the rule to produce more or fewer warnings?

Fewer warnings

### How will the change be implemented?

The rule no-wait-for-multiple-assertions currently forbids any multiple assertions inside a waitFor callback.
The proposal is to refine the rule so that it only reports an error when multiple assertions are made against the same asynchronous target.

This keeps the original intention (avoid useless timeouts when combining multiple checks on the same async call) while avoiding unnecessary errors in cases where separate async targets are being tested together.

### Example code

✅ Current rule reports (but should still report)
```ts
await waitFor(() => {
  expect(window.fetch).toHaveBeenCalledWith('/foo');
  expect(window.fetch).toHaveBeenCalledTimes(1);
});
```

Here, both assertions relate to the same async target (window.fetch).
If one of them is permanently failing, waitFor will unnecessarily retry until timeout.
The rule should still report this as a violation.

❌ Current rule reports (but should NOT report anymore)
```ts
await waitFor(() => {
  expect(window.fetch).toHaveBeenCalledWith('/foo');
  expect(localStorage.setItem).toHaveBeenCalledWith('bar', 'baz');
});
```

In this case, two different async side-effects are being checked (fetch and localStorage).
If both are eventually satisfied, waiting time is identical whether assertions are combined or separated.
If one fails permanently, the timeout will still happen regardless of being combined or separated.
So forcing developers to split them adds verbosity without clear benefit.

### How does the current rule affect the code?

- It flags all multiple assertions inside waitFor, even when they target different async behaviors.
- This often forces developers to artificially split assertions into multiple waitFor calls, which can hurt readability and lead to more verbose tests.

### How will the new rule affect the code?

- Multiple assertions against the same async target will still be forbidden.
- Multiple assertions against different async targets will be allowed.
- Developers will no longer need to split unrelated assertions into separate waitFor calls.

### Anything else?

There are also practical cases where multiple assertions in the same waitFor should be acceptable, because they are targeting different async targets and are expected to resolve independently. For example:

```tsx
export function Example() {
  const [a, setA] = useState('');
  const [b, setB] = useState('');

  useEffect(() => {
    setTimeout(() => setA('a'), 100);
    setTimeout(() => setB('b'), 500);
  }, []);

  return (
    <div>
      <div data-testid="a">{a}</div>
      <div data-testid="b">{b}</div>
    </div>
  );
}

it('Example', async () => {
  render(<Example />);

  await waitFor(() => {
    expect(screen.getByTestId('a')).toHaveTextContent('a');
    expect(screen.getByTestId('b')).toHaveTextContent('b');
  });
});

```

In this case, both `a` and `b` are independent async state updates.
Splitting them into two separate waitFor calls does not reduce the timeout risk (if one fails, it will still time out), but it does add verbosity.
Allowing multiple assertions on different async targets would make this kind of test more concise without sacrificing reliability.

### Do you want to submit a pull request to change the rule?

Yes

**Comment by Belco90** [comment]
I understand the motivation for this improvement, but the changes are quite complex for little benefit.

I can see how the _Common mistakes with RTL_ explains this scenario by asserting over the same mock, so the problem you described is correct (I probably didn't realize about this detail when created the rule).

However, how easy is to implement this detection? Additionally, the scenarios you mentioned seem fine, but are fairly simple

> The proposal is to refine the rule so that it only reports an error when multiple assertions are made against the same asynchronous target.

So what happens if multiple assertions against the same target are found, together with something else?

```js
await waitFor(() => {
  expect(window.fetch).toHaveBeenCalledTimes(1);
  expect(localStorage.setItem).toHaveBeenCalledWith('bar', 'baz');
  expect(window.fetch).toHaveBeenCalledWith('/foo'); // <--- this one is reported?
});
```

I'm sure there are other _controversial_ cases, but this is what I think off the top of my head. I'm not against implementing this improvement, but I'd like to discuss a bit further the benefits and the detection complexity.

**Comment by y-hsgw** [comment]
Thank you for your thoughtful input — I agree that this change introduces some complexity, and it's important to balance that with the actual benefits.

What I'm mainly aiming for is to **only report multiple expect() calls inside waitFor() when they target the same expression**.

That said, I’m starting to feel like detecting what counts as **“the same target”** might be trickier than I first thought 😅

Not sure, but maybe just comparing the first argument of each `expect()` using `getSourceCode(context).getText(arguments[0])` would be enough... 🤔


> So what happens if multiple assertions against the same target are found, together with something else?

In the case you mentioned, yes — I would still expect the rule to report the duplicated expect(window.fetch) call, even if it appears alongside unrelated assertions. Something like:
```ts
await waitFor(() => {
  expect(window.fetch).toHaveBeenCalledTimes(1);
  expect(localStorage.setItem).toHaveBeenCalledWith('bar', 'baz');
  expect(window.fetch).toHaveBeenCalledWith('/foo'); // ← this would be reported
});
```

**Comment by Belco90** [comment]
I see, LGTM! @y-hsgw I assume you'll work on this improvement.

@all-contributors please add @y-hsgw for ideas
